Sql editor compatibility + improved pattern geometry editing

landonreed opened the pull request:
This PR updates the UI to be compatible with the new SQL-based editor.  A list of changes includes:
- Converts a number of actions to use `redux-actions`
- GraphQL fetch for for all editor entities
- Change get/update/delete single entity actions to be generic (shared action logic that handles changing the endpoint and mapping for each entity type, e.g., `PUT /api/editor/${entityType}`
- Updates the pattern geometry editing logic:
  - Server now sends list of shape points
  - UI now handles projecting stops onto the pattern shape
  - Control points (anchor points) are persisted in the back end
  - Single, inter-stop (or inter-control point) segments are independently editable/visible to improve editing for loop routes or routes that double back on themselves.

landonreed committed:
use createActions

landonreed committed:
refactor validation error fetch

landonreed committed:
add detail to status modal

landonreed committed:
pass all props to status modal

landonreed committed:
refactor(editor): add mapbox routing code

landonreed committed:
style(user): add permissions warning if project is null

landonreed committed:
style(map): comment out log

landonreed committed:
fix(graphql): fix graphql reducers, more createAction updates

landonreed committed:
fix(status): handle create snapshot job finish

landonreed committed:
update validation ui

landonreed committed:
update fare type

landonreed committed:
style(status): clean up status reducer

landonreed committed:
refactor(editor): refactor/remove redundant actions

landonreed committed:
refactor(timetable): refactor timetable for sql backend

landonreed committed:
refactor(editor): refactor editor components for sql backend

landonreed committed:
feat(patterns): edit pattern geometry segment by segment

This update improves the pattern editing feature in the editor and adapts the UI to work with the
SQL backend. The backend represents the pattern geometry now as an array of shapepoints rather than
a simple GeoJSON linestring. This allows for saving metadata about the points, specifically whether
it is the projected point from a stop, a user-defined anchor point, or simple another coordinate in
the sequence. In the UI using selectors, we construct two arrays from these shape points: 1) an
array of control points to render for the user to manipulate and 2) an array of coordinate sets
(essentially a multi-linestring) frim which to render the geometry.

BREAKING CHANGE

landonreed committed:
build(package.json): update turf dependencies, add others

landonreed committed:
style(lint): fix lint

landonreed committed:
style(lint): fix lint

landonreed committed:
fix(editor): fix add stops mode

landonreed committed:
fix(editor): update sequence on pattern stop resequencing

landonreed committed:
style(flow): WIP flow fixes

landonreed committed:
fix(editor): recalculateSegment -> recalculateShape for tests

evansiroky commented on lib/editor/actions/map/stopStrategies.js:
```diff
@@ -1,16 +1,18 @@
+import along from '@turf/along'
 import ll from '@conveyal/lonlat'
 import lineDistance from 'turf-line-distance'
 import lineSlice from 'turf-line-slice'
```
We should try to use the latest turf packages

evansiroky commented on lib/common/actions/index.js:
```diff
@@ -1,37 +1,44 @@
 import { setErrorMessage } from '../../manager/actions/status'
 
-export function secureFetch (url, method = 'get', payload, raw) {
+export function secureFetch (url, method = 'get', payload, raw = false, isJSON = true) {
```
I'm seeing a lot of `res.json().then(` in other areas.  Perhaps we can add something that assumes a response is json or assume that if `raw` is false, then json should be returned.

evansiroky commented on lib/editor/actions/tripPattern.js:
```diff
@@ -1,196 +1,69 @@
+export const togglingPatternEditing = createAction('TOGGLE_PATTERN_EDITING')
+export const setActivePatternSegment = createAction('SET_ACTIVE_PATTERN_SEGMENT')
+const savedTripPattern = createAction('SAVED_TRIP_PATTERN')
+export const resnapStops = createAction('RESNAP_STOPS')
```
Really stoked to see all this work to convert to redux-actions.  👍

evansiroky commented on lib/editor/util/gtfs.js:
```diff
@@ -1,37 +1,88 @@
+  {id: 'agency', tableName: 'agency'}
 ]
+
+export function getTableById (tableData: any, id: string, emptyArrayOnNull: boolean = true): any {
```
I'm not quite sure why this is needed.  What is the reasoning behind this?

evansiroky commented on lib/editor/util/gtfs.js:
```diff
@@ -1,37 +1,88 @@
+// }
+
+// FIXME: Add flow.  Consolidate with gtfs/util/getEntityGraphQLRoot
+export const COMPONENT_LIST = [
```
Use an object for this if it is needed.  There'd be no need to do array searches.

evansiroky commented on lib/editor/components/EditorInput.js:
```diff
@@ -264,26 +264,35 @@ export default class EditorInput extends Component {
+        // below).
+        const stops = [...getTableById(tableData, 'stop')]
+        const stopIndex = stops.findIndex(s => s.id === activeEntity.id)
+        const selectedStop = stops.find(s => s.id === currentValue)
```
I'd recommend creating an index of stopsByStopId to speed this up a bit.

evansiroky commented on lib/editor/components/map/AddableStopsLayer.js:
```diff
@@ -23,13 +31,15 @@ export default class AddableStopsLayer extends Component {
             .filter(stop => {
               // filter out stops by lat/lng
-              if (stop.stop_lat > bounds.getNorth() || stop.stop_lat < bounds.getSouth() || stop.stop_lon > bounds.getEast() || stop.stop_lon < bounds.getWest()) {
+              if (this.isStopOutOfBounds(stop, bounds)) {
```
might be a good idea to put this in a selector

evansiroky commented on lib/editor/components/map/ControlPointsLayer.js:
```diff
@@ -1,69 +1,75 @@
+              ? 'blue'
+              : index === patternSegment + 1 // control point after
+              ? 'yellow'
+              : '#888' // inactive
```
Good idea to make this visual distinction.

evansiroky commented on lib/editor/components/map/DirectionIconsLayer.js:
```diff
@@ -4,43 +4,67 @@ import { divIcon } from 'leaflet'
+          position.geometry.coordinates[1] > bounds.getNorth() ||
+          position.geometry.coordinates[1] < bounds.getSouth() ||
+          position.geometry.coordinates[0] > bounds.getEast() ||
+          position.geometry.coordinates[0] < bounds.getWest()
```
Might be good to make a util function for whether a point is within a bounds object.

evansiroky replied:
Better yet, use [bounds.contains](http://leafletjs.com/reference-1.3.0.html#bounds-contains)

evansiroky commented on lib/editor/components/map/PatternsLayer.js:
```diff
@@ -62,25 +67,143 @@ export default class PatternsLayer extends Component {
   }
 }
+
+class PatternSegment extends Component {
```
Consider extracting this class into a new file.

evansiroky commented on lib/editor/containers/ActiveTripPatternList.js:
```diff
@@ -13,67 +13,80 @@ import {
+  setErrorMessage,
+  togglePatternEditing,
+  undoActiveTripPatternEdits,
+  updatePatternGeometry
```
🥇

evansiroky commented on lib/editor/util/map.js:
```diff
@@ -226,230 +240,356 @@ export async function recalculateShape ({
+      shapePoints.push(...coordinatesToShapePoints(segment, lastDistance, shapePoints.length))
     }
   }
+  const resquencedShapePoints = shapePoints.map(resequenceShapePoints)
```
`resquencedShapePoints` is a typo

landonreed committed:
style(logs): remove log statement

landonreed committed:
fix(editor): refactor generic actions and fix feed info

landonreed committed:
refactor(manager): refactor projects reducer/actions

landonreed committed:
refactor(status): remove unused action

landonreed committed:
refactor(manager): refactor actions/reducers

landonreed committed:
refactor(manager): simplify containers' use of actions + formatting

landonreed committed:
fix(project): fix bounding box project settings form

landonreed committed:
fix(routing): Use GraphHopper for follow streets routing

Mapzen (and thus the valhalla routing service) is no longer in operation. This addresses #60, but
needs to be cherry-picked into a separate branch for merging into dev.

refs #60

landonreed committed:
fix(editor): fix avg speed travel time calculation for pattern stops

landonreed committed:
fix(timetable): check that a time cell isn't a frequency start/end time

This fix avoids an NPE when attempting to fetch the pattern stop with an undefined stopTimeIdx.
Frequency entries in the timetable do not need stop time objects.

landonreed committed:
fix(editor): misc. editor fixes for SQL editor

landonreed committed:
docs(docs): add api interaction instructions

Commit:
make first sentence a sentence

landonreed committed:
fix(editor): fix schedule exceptions for sql editor

landonreed committed:
fix(editor): fix frequency trip editing for SQL editor

landonreed committed:
feat(editor): restrict editing for a feed to a single session

This adds a procedure to "check out" a feed for editing. If the feed is already being edited (by
another user or in another tab), the user is prevented from editing it. When a feed is checked out
for editing, the UI will make calls to the server every ten seconds to indicate that the feed is
still checked out. When a user exits the Editor or closed the browser tab/window, a request is made
to the server to remove the lock on the feed.

landonreed committed:
fix(validation): fix check for fatal exception on validation result

landonreed committed:
add fixme

Commit:
add instructions to fetch new feed version by URL

landonreed committed:
fix(messages): fix deployment settings messages

landonreed committed:
feat(status): add reload action to status modal

landonreed committed:
perf(logging): remove loud logging statement

landonreed committed:
fix(editor lock): fix locking functionality

landonreed committed:
fix(editor): fix start from scratch

landonreed committed:
fix(deployments): fix alignment of button dropdown

landonreed committed:
fix(misc): misc sql-editor bug fixes

landonreed committed:
fix(editor lock): clear GTFS content after lock removal

landonreed committed:
store feed ID on fetch to provide feed param to lock

landonreed committed:
fix(editor): fix message on first snapshot creation

landonreed committed:
fix(editor): fix trip writes for non-frequency based trips

landonreed committed:
fix(editor): fix new calendar writes

landonreed committed:
fix(editor lock): add session id query param to write operations

landonreed committed:
fix(flow): update control point flow def

landonreed committed:
fix(editor): update store when gtfs entity saved successfully

landonreed committed:
refactor(snapshots): simplify snapshot actions

landonreed committed:
fix(editor): add refetch option to create GTFS entity method

landonreed committed:
docs(editor): add comments for map methods

landonreed committed:
fix(editor): clear editor state on component mount

landonreed committed:
fix(patterns): fix pattern stop removal segment splice

landonreed committed:
fix(patterns): log warning if pattern stop not draggable

landonreed committed:
fix(patterns): fix reverse pattern and calculate default travel time

landonreed committed:
fix(editor): fix fare transfers default value

landonreed committed:
fix(editor): fetch only required fields for feed info

landonreed committed:
fix(editor): export receiveFeedSource action

landonreed committed:
fix(status): fix handling of completed snapshot-related jobs

landonreed committed:
fix(deployment): disable feed deployment if no versions exist

landonreed committed:
fix(manager): fix on load feed version for editing

landonreed committed:
fix(r5): handle r5 build network job correctly

landonreed committed:
fix(editor): fix shape point snapping

landonreed committed:
fix(patterns): ensure that control point drag ends in correct loc.

landonreed committed:
fix(patterns): fix pattern geometry update on anchor drag

landonreed committed:
style(lint): fix lint and flow

landonreed committed:
fix(jest): temporarily disable mapzen pattern shape tests

landonreed committed:
ci(travis): install urllib3 to avoid SSL insecure warnings

landonreed committed:
ci(travis): update python version to fix ssl issue

landonreed committed:
ci(travis): use python 3.3.3

landonreed committed:
ci(travis): use python 2.7.14

landonreed committed:
fix(import): fix capitalization for turf-linestring

landonreed committed:
fix(import): fix capitalization for turf-linestring

landonreed committed:
fix(gtfs): fix wheelchair boarding values

landonreed committed:
fix(editor): formatting + fix monday selection in exceptions

landonreed committed:
style(editor): remove commented out code

landonreed committed:
style(prettier): format file with prettier

landonreed committed:
fix(editor): fix entity type checks using field typeof

landonreed committed:
fix(editor): fix timepoint checkbox

landonreed committed:
fix(editor): fix route#wheelchair_accessible field name/values

landonreed committed:
fix(manager): use file-saver for file downloads

fixes #65

landonreed committed:
fix(editor): fix schedule exceptions swap type

fixes #74

landonreed committed:
style(lint): fix indentation; remove log statement

landonreed committed:
fix(editor): fix view all route alignments on editor map

fixes #76

landonreed committed:
style(lint): fix lint

landonreed committed:
refactor(graphql): refactor graphql requests to match backend update

The GTFS GraphQL endpoint in datatools-server was updated to conform to the GraphQL specification in
catalogueglobal/datatools-server#76. This commit provides the corresponding changes in the client.

landonreed committed:
style(lint): fix whitespace and camel_case var names

landonreed committed:
style(log): remove dev log statement

landonreed committed:
fix(manager): fix note count update for feed source and versions

fixes #77

landonreed committed:
fix(editor): fixes "scheduleexception" misspelling and create stop bug

fixes #79, fixes #78

landonreed committed:
fix(editor): fix pattern direction icons/text

fixes #80

landonreed committed:
fix(editor): fix "scheduleexception" text in delete/duplicate button

refs #79

landonreed committed:
perf(editor): use shouldComponentUpdate for virtualized select and store options in state

fixes #75

landonreed committed:
fix(editor): reset trip pattern edits after error on save

landonreed committed:
fix(editor): request zone_id with stops table

fixes #100

landonreed committed:
fix(editor): fix stop/route select input

fixes #92

landonreed committed:
fix(snapshot): refetch base GTFS if namespace changes (when snapshot restored)

fixes #101

landonreed committed:
style(lint): fix whitespace on ternaries

landonreed committed:
fix(editor): update pattern geometry when shape is deleted

fixes #109

Commit:
feat(common): Clean up JobMonitor display and functionality. Addresses #85

landonreed committed:
fix(editor): handle travel time calc for patterns without shapes

fixes #107

Commit:
feat(snapshots): Remove concept of "active" snapshot from UI. Fixes #90, #102

landonreed committed:
fix(timetable): add pattern/route/calendar grouped trip counts

fixes #110

Commit:
fix(manager): Fix error with FeedVersion viewer and map. Should fix #86/#87/#88

landonreed committed:
fix(editor): prevent reversing pattern if pattern has trips

fixes #104

landonreed committed:
fix(patterns): fix add stop at click and interval; refactor utils

fixes #108

landonreed committed:
style(lint): fix lint

landonreed committed:
style(lint): fix lint

landonreed committed:
style(lint): fix lint/formatting

Commit:
fix(feed viewer): Add limit=-1 to show all routes/stops/patterns/shape pts

Commit:
fix(validation viewer): Don't show "no validation errors" until we actually get a response

Commit:
feat(feed viewer): Add various enhancements to feed viewer, including consistency in selected route/

landonreed committed:
refactor(editor): refactor schedule exception component

landonreed committed:
fix(editor): clone exceptions to avoid mutation of store

fixes #103

landonreed committed:
fix(editor): fix feed info reset/undo

fixes #97

landonreed committed:
perf(temporarily disable entity selector if list > 10,000):

refs #75

landonreed committed:
style(editor): misc formatting, white space, comments

landonreed committed:
fix(editor): handle bad entity and pattern IDs in editor URL

fixes #99

landonreed committed:
refactor(editor): add missing args for add stops at interval

refs #108

landonreed committed:
refactor(timetable): add stop ID to stop time column headers

landonreed committed:
fix(editor): default new stop location to center of bounds (rather than 0,0)

landonreed committed:
style(manager): lint fixes

landonreed committed:
refactor(editor): simplify data reducer, remove log statements, general clean up

landonreed committed:
fix(editor): fix undo for fields with null values

fixes #113

landonreed committed:
style(lint): remove unused import

Commit:
fix(feed viewer): Reset Feed Viewer to "summary" tab when FeedVersion changes

landonreed commented:
After extensive testing, I am merging this branch into dev.
